core/trino-main/src/main/java/io/trino/metadata/HandleResolver.java

```
@@ -13,106 +13,66 @@
     }
 
-    public synchronized void removeCatalogHandleClasses(String catalogName)
+    public void unregisterClassLoader(PluginClassLoader classLoader)
     {
-        Set<Class<?>> classes = catalogHandleClasses.remove(catalogName);
-        checkState(classes != null, "Catalog not registered: %s", catalogName);
-        for (Class<?> handleClass : classes) {
-            if (allHandleClasses.remove(handleClass, 1) == 1) {
-                handleClassIds.remove(handleClass);
-            }
-        }
+        boolean result = classLoaders.remove(classLoader.getId(), classLoader);
+        checkState(result, "Class loader not registered: " + classLoader.getId());
     }
 
-    public synchronized String getId(Object tableHandle)
+    @SuppressWarnings("MethodMayBeStatic")
+    public String getId(Object tableHandle)
     {
         Class<?> handleClass = tableHandle.getClass();
-        String id = handleClassIds.get(handleClass);
-        checkArgument(id != null, "Handle class not registered: " + handleClass.getName());
-        return id;
+        return classId(handleClass);
     }
 
-    public synchronized Class<?> getHandleClass(String id)
+    public Class<?> getHandleClass(String id)
     {
-        Class<?> handleClass = handleClassIds.inverse().get(id);
-        checkArgument(handleClass != null, "Handle ID not found: " + id);
-        return handleClass;
+        int splitPoint = id.lastIndexOf(':');
+        checkArgument(splitPoint > 1, "Invalid handle id: " + id);
+        String classLoaderId = id.substring(0, splitPoint);
+        String className = id.substring(splitPoint + 1);
+
+        ClassLoader classLoader = classLoaders.get(classLoaderId);
+        checkArgument(classLoader != null, "Unknown handle id: " + id);
+
+        try {
+            Class<?> handleClass = classLoader.loadClass(className);
+            if (handleClass != null) {
```

Is this possible? It should throw instead of returning null

---

It is not possible, but the old code was checking... I'll remove it